src/System.Security.Cryptography.ProtectedData/src/System/Security/Cryptography/ProtectedData.cs

```diff
@@ -0,0 +1,106 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+using Internal.Cryptography;
+
+using DATA_BLOB = Interop.Crypt32.DATA_BLOB;
+using CryptProtectDataFlags = Interop.Crypt32.CryptProtectDataFlags;
+
+namespace System.Security.Cryptography
+{
+    public static partial class ProtectedData
+    {
+        public static byte[] Protect(byte[] userData, byte[] optionalEntropy, DataProtectionScope scope)
+        {
+            if (userData == null)
+                throw new ArgumentNullException("userData");
+
+            return ProtectOrUnprotect(userData, optionalEntropy, scope, protect: true);
+        }
+
+        public static byte[] Unprotect(byte[] encryptedData, byte[] optionalEntropy, DataProtectionScope scope)
+        {
+            if (encryptedData == null)
+                throw new ArgumentNullException("encryptedData");
+
+            return ProtectOrUnprotect(encryptedData, optionalEntropy, scope, protect: false);
+        }
+
+        private static byte[] ProtectOrUnprotect(byte[] inputData, byte[] optionalEntropy, DataProtectionScope scope, bool protect)
+        {
+            Debug.Assert(inputData != null);
+
+            unsafe
+            {
+                fixed (byte* pInputData = inputData, pOptionalEntropy = optionalEntropy)
+                {
+                    DATA_BLOB userDataBlob = new DATA_BLOB((IntPtr)pInputData, (uint)(inputData.Length));
+                    DATA_BLOB optionalEntropyBlob;
+                    if (optionalEntropy != null)
+                    {
+                        optionalEntropyBlob = new DATA_BLOB((IntPtr)pOptionalEntropy, (uint)(optionalEntropy.Length));
+                    }
+
+                    // For desktop compat, we ignore unknown bits in the "scope" value rather than throwing.
+                    CryptProtectDataFlags flags = CryptProtectDataFlags.CRYPTPROTECT_UI_FORBIDDEN;
+                    if (scope == DataProtectionScope.LocalMachine)
+                    {
+                        flags |= CryptProtectDataFlags.CRYPTPROTECT_LOCAL_MACHINE;
+                    }
+
+                    DATA_BLOB outputBlob = default(DATA_BLOB);
+                    try
+                    {
+                        bool success = protect ?
+                            Interop.Crypt32.CryptProtectData(&userDataBlob, null, &optionalEntropyBlob, IntPtr.Zero, IntPtr.Zero, flags, &outputBlob) :
```

Comment:
A) Can we use ref instead of the & operator here?
B) What is being sent for optionalEntropyBlob when optionalEntropy is null?  I didn't know that one could do address-of an uninitialized local...